### PR TITLE
add montecarlo integrate api

### PR DESCRIPTION
**Pull request by cxxly**
### PR types

New fatures
### PR changes

APIs
### Describe

Add montecarlo integration.

### CODE REVIEW

File: ppsci/experimental/math_module.py [hunk 1]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
         return paddle.cumulative_trapezoid(y, x, dx, axis)
     else:
         raise ValueError(f'mode should be "sum" or "cumsum", but got {mode}')
+
+
+def montecarlo_integrate(
+    fn: callable,
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:469]:
```suggestion
    fn: Callable,
```

File: ppsci/experimental/math_module.py [hunk 2]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
...
+    dim: int,
+    N: int = 1000,
+    integration_domain: Optional[List[List[float]], paddle.Tensor] = None,
+    seed: int = None,
+):
+    """Integrates the passed function on the passed domain using vanilla Monte
+    Carlo Integration.
+
+    Args:
+        fn (func): The function to integrate over.
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:479]:
```suggestion
        fn (Callable): The function to integrate over.
```

File: ppsci/experimental/math_module.py [hunk 3]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
...
+        dim (int): Dimensionality of the function's domain over which to
+            integrate.
+        N (int, optional): Number of sample points to use for the integration.
+            Defaults to 1000.
+        integration_domain (list or backend tensor, optional): Integration
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:484]:
```suggestion
        integration_domain (Optional[List[List[float]], paddle.Tensor]): Integration
```

File: ppsci/experimental/math_module.py [hunk 4]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
...
+        ValueError: If len(integration_domain) != dim
+
+    Returns:
+        Integral value
+
+    Examples:
+        >>> import paddle
+        >>> import ppsci
+
+        >>> paddle.seed(1024)
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:500]:
```suggestion
        >>> _ = paddle.seed(1024)
```

File: ppsci/experimental/math_module.py [hunk 5]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
...
+        Tensor(shape=[], dtype=float32, place=Place(gpu:0), stop_gradient=True, 3.25152588)
+    """
+
+    @expand_func_values_and_squeeze_integral
+    def calculate_result(function_values, integration_domain):
+        """Calculate an integral result from the function evaluations
+
+        Args:
+            function_values (backend tensor): Output of the integrand
+            integration_domain (backend tensor): Integration domain
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:528]: 函数注释里所有backend tensor改为paddle.Tensor

File: ppsci/experimental/math_module.py [hunk 6]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
...
+        scales = integration_domain[:, 1] - integration_domain[:, 0]
+        volume = paddle.prod(scales)
+
+        # Integral = V / N * sum(func values)
+        N = function_values.shape[0]
+        integral = volume * paddle.sum(function_values, axis=0) / N
+        return integral
+
+    def calculate_sample_points(
+        N: int, integration_domain: paddle.Tensor, seed: int = None
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:542]:
```suggestion
        N: int, integration_domain: paddle.Tensor, seed: Optional[int] = None
```

File: ppsci/experimental/math_module.py [hunk 7]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
...
+    ):
+        """Calculate random points for the integrand evaluation
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:544]: 句子结尾加上句号.

File: ppsci/experimental/math_module.py [hunk 8]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
...
+
+        Args:
+            N (int): Number of points
+            integration_domain (backend tensor): Integration domain
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:548]: 句子结尾加上句号.

File: ppsci/experimental/math_module.py [hunk 9]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
...
+
+
+def _setup_integration_domain(
+    dim: int, integration_domain: Optional[List[List[float]], paddle.Tensor]
+) -> paddle.Tensor:
+    """Sets up the integration domain if unspecified by the user.
+    Args:
+        dim (int): Dimensionality of the integration domain.
+        integration_domain (List or Tensor): Integration domain, e.g. [[-1,1],[0,1]]. Defaults to [-1,1]^dim.
+    Returns:
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:589]:
```suggestion

    Returns:
```

File: ppsci/experimental/math_module.py [hunk 10]
```diff
@@ -462,3 +463,183 @@ def trapezoid_integrate(
...
+
+    if tuple(integration_domain.shape) != (dim, 2):
+        raise ValueError(
+            "The integration domain has an unexpected shape. "
+            f"Expected {(dim, 2)}, got {integration_domain.shape}"
+        )
+    return integration_domain
+
+
+def evaluate_integrand(fn, points, weights=None, args=None):
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:608]: 该函数为非公开的辅助函数，前面加上单个下划线
```suggestion
def _evaluate_integrand(fn, points, weights=None, args=None):
```

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

File: ppsci/experimental/math_module.py (hunk 1 above)
Comment (cxxly) [reply ppsci/experimental/math_module.py:469]: DONE

File: ppsci/experimental/math_module.py (hunk 2 above)
Comment (cxxly) [reply ppsci/experimental/math_module.py:479]: DONE

File: ppsci/experimental/math_module.py (hunk 3 above)
Comment (cxxly) [reply ppsci/experimental/math_module.py:484]: DONE

File: ppsci/experimental/math_module.py (hunk 5 above)
Comment (cxxly) [reply ppsci/experimental/math_module.py:528]: DONE

File: ppsci/experimental/math_module.py (hunk 6 above)
Comment (cxxly) [reply ppsci/experimental/math_module.py:542]: DONE

File: ppsci/experimental/math_module.py (hunk 8 above)
Comment (cxxly) [reply ppsci/experimental/math_module.py:548]: DONE

File: ppsci/experimental/math_module.py (hunk 7 above)
Comment (cxxly) [reply ppsci/experimental/math_module.py:544]: DONE

File: ppsci/experimental/math_module.py (hunk 10 above)
Comment (cxxly) [reply ppsci/experimental/math_module.py:608]: DONE

File: ppsci/experimental/math_module.py [hunk 11]
```diff
@@ -462,3 +463,182 @@ def trapezoid_integrate(
         return paddle.cumulative_trapezoid(y, x, dx, axis)
     else:
         raise ValueError(f'mode should be "sum" or "cumsum", but got {mode}')
+
+
+def montecarlo_integrate(
+    fn: Callable,
+    dim: int,
+    N: int = 1000,
+    integration_domain: Optional[List[List[float]], paddle.Tensor] = None,
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:472]:
```suggestion
    integration_domain: Union[List[List[float]], paddle.Tensor] = None,
```

File: ppsci/experimental/math_module.py [hunk 12]
```diff
@@ -462,3 +463,182 @@ def trapezoid_integrate(
...
+    """Integrates the passed function on the passed domain using vanilla Monte
+    Carlo Integration.
+
+    Args:
+        fn (Callable): The function to integrate over.
+        dim (int): Dimensionality of the function's domain over which to
+            integrate.
+        N (Optional[int]): Number of sample points to use for the integration.
+            Defaults to 1000.
+        integration_domain (Optional[List[List[float]], paddle.Tensor]): Integration
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:484]:
```suggestion
        integration_domain (Union[List[List[float]], paddle.Tensor]): Integration
```

File: ppsci/experimental/math_module.py [hunk 13]
```diff
@@ -462,3 +463,182 @@ def trapezoid_integrate(
...
+                raise ValueError("N has to be a positive integer.")
+
+    integration_domain = _setup_integration_domain(dim, integration_domain)
+    sample_points = calculate_sample_points(N, integration_domain, seed)
+    function_values, _ = _evaluate_integrand(fn, sample_points)
+    return calculate_result(function_values, integration_domain)
+
+
+def _setup_integration_domain(
+    dim: int, integration_domain: Optional[List[List[float]], paddle.Tensor]
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:581]:
```suggestion
    dim: int, integration_domain: Union[List[List[float]], paddle.Tensor]
```

File: ppsci/experimental/math_module.py [hunk 14]
```diff
@@ -462,3 +463,182 @@ def trapezoid_integrate(
...
+        >>> # Compute the function integral by sampling 10000 points over domain
+        >>> integral_value = ppsci.experimental.montecarlo_integrate(
+        >>>    some_function,
+        >>>    dim=2,
+        >>>    N=10000,
+        >>>    integration_domain=[[0, 1], [-1, 1]]
+        >>> )
+
+        >>> print(integral_value)
+        Tensor(shape=[], dtype=float32, place=Place(gpu:0), stop_gradient=True, 3.25152588)
```
Comment (HydrogenSulfate) [inline ppsci/experimental/math_module.py:517]:
```suggestion
        >>> import paddle
        >>> import ppsci

        >>> _ = paddle.seed(1024)
        >>> # The function we want to integrate, in this example
        >>> # f(x0,x1) = sin(x0) + e^x1 for x0=[0,1] and x1=[-1,1]
        >>> # Note that the function needs to support multiple evaluations at once (first
        >>> # dimension of x here)
        >>> # Expected result here is ~3.2698
        >>> def some_function(x):
        ...     return paddle.sin(x[:, 0]) + paddle.exp(x[:, 1])

        >>> # Compute the function integral by sampling 10000 points over domain
        >>> integral_value = ppsci.experimental.montecarlo_integrate(
        ...     some_function,
        ...     dim=2,
        ...     N=10000,
        ...     integration_domain=[[0, 1], [-1, 1]],
        ... )

        >>> print(integral_value)
        Tensor(shape=[], dtype=float32, place=Place(gpu:0), stop_gradient=True,
               3.25152588)
```

**Review by HydrogenSulfate — APPROVED** [review]
LGTM

**Pull request merged**